presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ProjectionPushThroughJoin.java

```diff
@@ -0,0 +1,246 @@
+                boolean pushdownAll = true;
+                ImmutableMap.Builder<Symbol, Expression> assignmentBuilder = ImmutableMap.builder();
+                for (Map.Entry<Symbol, Expression> entry : node.getAssignments().entrySet()) {
+                    Symbol key = entry.getKey();
+                    Expression expression = entry.getValue();
+                    if (expression instanceof SymbolReference) {
+                        if (((SymbolReference) expression).equals(key.toSymbolReference())) {
+                            assignmentBuilder.put(key, expression);
+                            continue;
+                        }
+                    }
+                    if (isPushDownExpression(expression)) {
+                        Set<Symbol> fields = DependencyExtractor.extractUnique(expression);
+                        if (fields.isEmpty()) {
+                            assignmentBuilder.put(key, expression);
+                            continue;
+                        }
+                        if (leftOutputSymbols.containsAll(fields) && !rightOutputSymbols.stream().anyMatch(s -> fields.contains(s))) {
+                            assignmentBuilder.put(key, new SymbolReference(key.getName()));
+                            leftAssignments.put(key, expression);
+                            continue;
+                        }
+                        if (rightOutputSymbols.containsAll(fields) && !leftOutputSymbols.stream().anyMatch(s -> fields.contains(s))) {
+                            assignmentBuilder.put(key, new SymbolReference(key.getName()));
+                            rightAssignments.put(key, expression);
+                            continue;
+                        }
+                    }
+                    pushdownAll = false;
+                    assignmentBuilder.put(key, expression);
+                }
+
+                if (!leftAssignments.isEmpty() || !rightAssignments.isEmpty()) {
+                    if (pushdownAll) {
+                        return context.rewrite(node.getSource(), new PushdownContext(leftAssignments, rightAssignments));
+                    }
+                    return new ProjectNode(idAllocator.getNextId(), context.rewrite(node.getSource(), new PushdownContext(leftAssignments, rightAssignments)), assignmentBuilder.build());
+                }
+                return context.defaultRewrite(node);
+            }
+            return buildNode(node, node.getSource().getOutputSymbols(), context);
+        }
+
+        @Override
+        public PlanNode visitJoin(JoinNode node, RewriteContext<PushdownContext> context)
+        {
+            PushdownContext assignment = context.get();
+            if (assignment == null) {
+                return context.defaultRewrite(node);
+            }
+
+            PlanNode left = node.getLeft();
+            PlanNode right = node.getRight();
+            Map<Symbol, Expression> leftAssignments = assignment.getLeftAssignment();
```

Comment:
I think context should have all projections without separation into left and right set. In this function you should split assignments into those that can be pushed left and those that can be pushed right.
What do you think?